libs/gltfio/src/extended/TangentSpaceMeshWrapper.cpp

```diff
@@ -0,0 +1,330 @@
+        std::memcpy(data, (T) mPassthrough.data[static_cast<int>(attribute)], nbytes);
+    }
+
+    size_t getVertexCount() const noexcept { return mPassthrough.vertexCount; }
+    size_t getTriangleCount() const noexcept { return mPassthrough.triangleCount; }
+
+private:
+    Passthrough mPassthrough;
+};
+
+// Note that the method signatures here match TangentSpaceMesh::Builder
+struct PassthroughBuilder {
+    void vertexCount(size_t count) noexcept { mPassthrough.vertexCount = count; }
+
+    void normals(float3 const* normals) noexcept {
+        mPassthrough.data[Passthrough::NORMALS] = normals;
+    }
+
+    void tangents(float4 const* tangents) noexcept {
+        mPassthrough.data[Passthrough::TANGENTS] = tangents;
+    }
+    void uvs(float2 const* uvs) noexcept { mPassthrough.data[Passthrough::UV0] = uvs; }
+
+    void positions(float3 const* positions) noexcept {
+        mPassthrough.data[Passthrough::POSITION] = positions;
+    }
+
+    void triangleCount(size_t triangleCount) noexcept {
+        mPassthrough.triangleCount = triangleCount;
+    }
+
+    void triangles(uint3 const* triangles) noexcept {
+        mPassthrough.data[Passthrough::TRIANGLES] = triangles;
+    }
+
+    void aux(AuxType type, void* data) noexcept {
+        mPassthrough.data[static_cast<int>(type)] = data;
+    }
+
+    PassthroughMesh* build() const noexcept {
+        return new PassthroughMesh(mPassthrough);
+    }
+
+private:
+    Passthrough mPassthrough;
+    friend struct PassthroughMesh;
+};
+
+} // anonymous
+
+#define DO_MESH_IMPL(METHOD, ...)                                                                  \
+    do {                                                                                           \
+        if (mTsMesh) {                                                                             \
+            mTsMesh->METHOD(__VA_ARGS__);                                                          \
+        } else {                                                                                   \
+            mPassthroughMesh->METHOD(__VA_ARGS__);                                                 \
+        }                                                                                          \
+    } while (0)
+
+#define DO_MESH_RET_IMPL(METHOD)                                                                   \
+    do {                                                                                           \
+        if (mTsMesh) {                                                                             \
+            return mTsMesh->METHOD();                                                              \
+        } else {                                                                                   \
+            return mPassthroughMesh->METHOD();                                                     \
+        }                                                                                          \
+    } while (0)
```

Comment:
is this to avoid using the interface? just curious.

Reply:
Yeah, just to avoid interfaces and virtual calls. But maybe it's ok here since it's on being called every frame or anything like that.